### ISSUE TITLE
Generate .DEF file for one module with several exports

### ISSUE THREAD

**Issue by moteus** [keplerproject/luarocks#284]
For example luasec library contain ssl.lua and ssl.dll

``` Lua
-- ssl.lua
local core    = require("ssl.core")
local context = require("ssl.context")
local x509    = require("ssl.x509")
...
```

So luarocks should install ssl.dll that exports `luaopen_ssl_core`, `luaopen_ssl_context` and `luaopen_ssl_x509`.
But luarocks create def file for `luaopen_ssl` and fails because there no such function.
And this is fragment of rockspec

``` Lua
...
install = {
  lib = {ssl = "ssl.dll"},
  lua = {"src/ssl.lua", ['ssl.https'] = "src/https.lua"}
},
modules = {
  ssl = { ... compile options ...}
}
...
```

And some times it is not possible split module to several. In fact for this case we even no need def file at all because luasec has defenition `LSEC_API=__declspec(dllexport)` for windows that works for MSVC and MinGW.

**Comment by ignacio** [comment]
The module writer should make sure the luaopen_foo_blah functions are properly exported.
But, as you mention, you might need to paper over an existing module you have no (easy) way of modifying (like applying patches specified in the rockspec).

So, the rockspec could specify a list of names of functions to be forcefully exported when building.
With MSVC compilers, the should be added to the .def file. With mingw, @Tieske knows better.
With gcc, everything is visible by default. So I wouldn't worry much about that.

This would imply a change to the rockspec format, so I don't know if it feasible to introduce in LR 2.x  ( @hishamhm ?)

Currently, the only way I can think of is to apply a patch within the rockspec.

**Comment by Tieske** [comment]
I'm not sure I understand  the issue, but the MinGW build works, as per rockspec in https://github.com/brunoos/luasec/pull/26 

@moteus can you give that rockspec a try with the MS compilers?

**Comment by moteus** [comment]
Problem is that dll should export `luaopen_ssl_core`, `luaopen_ssl_context` etc. 
But luarocks make .def file while building using MSVC compiler with function `luaopen_ssl`.
But `ssl.dll` has not such function and build fail because linker can not find export function.

The problem is that some modules do not properly exports their simbols (`__declspec(dllexport)`) for MS compiler and trick with def file allows build them on MS compiler. But in this case module correctly exports its function but it can be build because of luarocks.

**Comment by ignacio** [comment]
Yes, exactly. This particular rock now works, but some others do not. For instance, imagine a module comprised of two parts:
foo.core (which is a C library) and foo.lua

The current mechanism will generate a .def file exporting luaopen_foo, having no way of defining luaopen_foo_core. And, in this case, it gives an error when linking because luaopen_foo does not exist.

**Comment by hishamhm** [comment]
@ignacio, in that scenario, what is the filename of the "foo.core" module called? I'm not sure I follow the example. If a rockspec has, say:

```lua
build = {
   type = "builtin",
   modules = {
      ["foo.core"] = "foo/core.c",
      ["foo"] = "foo.lua",
   }
}
```

doesn't it properly create a foo/core.dll file with a luaopen_foo_core export?

**Comment by moteus** [comment]
in luasec you have ssl.dll which exports several luaopen_ssl_xxx functions and ssl.lua which uses this dll.

**Comment by hishamhm** [comment]
@moteus, I understand. I was asking @ignacio about the slightly different scenario he described (if I understood correctly).

For the scenario you described (situations where a single dll includes multiple Lua modules but the source code does not export them properly (for example, if LuaSec didn't have the LSEC_API option) ), the only way in LuaRocks 2.x, as @ignacio suggested, is to include a patch in the Makefile to add the exports to the source code.

**Comment by ignacio** [comment]
@hishamhm Yes, but I was expecting foo.dll to be generated instead of core.dll
I usually write some modules as a couple of files. A high level part in Lua and a low level one in C, and name each the same, so I know they belong together.
For instance:
- lualiblogger.lua
- lualiblogger.dll

(I found that avoids confusion for some users, if I ask them to look for a libary, to have a specific name instead of the generic "core").

**Comment by hishamhm** [comment]
ah, well... I guess that's less confusing when users are installing by hand and actually deal with the .dll and .lua files, but if you're telling them to just `luarocks install foo` and not care about what/where the files are, as long as `require("foo")` works, then I guess foo/core.dll is not too bad? :-)

**Comment by ignacio** [comment]
Ah, yes. In that case it is not that bad. The only thing that still stands is multiple exports (not my example, but a valid point raised by @moteus in the general case).

**Comment by hishamhm** [comment]
Another addition for LuaRocks 3 then? It's good that we start collecting the stuff we want to change/add.

**Comment by ignacio** [comment]
:+1: to that.

On a related note, I knew that we had a similar conversation [ages ago](http://lua-users.org/lists/lua-l/2009-12/msg00116.html) :smile:

**Comment by hishamhm** [comment]
Oh boy, and even then I was already worried about backward compatibility (with LR1, no less!) :D
